Sources/Conduit/Auth/OAuth2RequestPipelineMiddleware.swift

```diff
@@ -72,19 +72,23 @@ public struct OAuth2RequestPipelineMiddleware: RequestPipelineMiddleware {
             tokenStorage.lockRefreshToken(timeout: tokenRefreshLockRelinquishInterval, client: clientConfiguration, authorization: authorization)
             logger.info("Token is expired, proceeding to refresh token")
             OAuth2TokenRefreshCoordinator.shared.beginTokenRefresh()
+            
             refresh(token: token) { result in
-                self.tokenStorage.unlockRefreshTokenFor(client: self.clientConfiguration, authorization: self.authorization)
                 switch result {
                 case .error(let error):
                     logger.warn("There was an error refreshing the token")
                     if case OAuth2Error.clientFailure = error {
                         self.tokenStorage.removeTokenFor(client: self.clientConfiguration, authorization: self.authorization)
                     }
+                    tokenStorage.unlockRefreshTokenFor(client: clientConfiguration, authorization: authorization)
                     completion(.error(error))
                 case .value(let newToken):
                     logger.info("Successfully refreshed token")
                     logger.debug("New token issued: \(newToken)")
                     self.tokenStorage.store(token: newToken, for: self.clientConfiguration, with: self.authorization)
+                    tokenStorage.unlockRefreshTokenFor(client: clientConfiguration, authorization: authorization)
+
+                    /// Resume original request using freshly obtained bearer token
                     self.makeRequestByApplyingAuthorizationHeader(to: request, with: newToken, completion: completion)
                 }
```

Comment:
With the switch being exhaustive, what real difference is it making to but the unlock into each case? Isn't it happening 100% of the time anyways? I do see that technically each case has certain things happen before the unlock. Is that the important thing with this change?

Reply:
That's correct, the unlock action itself does (and must) still occur 100% of the time. The importance is more so _when_ the unlock occurs.  This change ensures that the token store has unquestionably handled the outcome of the token refresh (i.e. storing the new token or deleting the existing) _before_ opening the token store back up and resuming normal usage of it